kernel/src/process_loading.rs

```diff
@@ -883,11 +1285,136 @@ impl<'a, C: Chip, D: ProcessStandardDebug> ProcessLoadingAsync<'a>
+                let (padding_requirement, previous_app_end_addr, next_app_start_addr) =
+                    (pr, prev_app_addr, next_app_addr);
+                if let Some(mut metadata) = self.process_binaries_metadata.get() {
+                    metadata.new_app_address = app_address;
+                    metadata.new_app_size = new_app_size;
+                    self.process_binaries_metadata.set(metadata);
+                }
+                Ok((
+                    app_address,
+                    padding_requirement,
+                    previous_app_end_addr,
+                    next_app_start_addr,
+                ))
+            }
+            Err(e) => Err(e),
+        }
+    }
+
+    fn check_new_binary_validity(
+        &self,
+        new_app_address: Option<usize>,
+    ) -> Result<(), ProcessBinaryError> {
+        let flash = self.flash_bank.get();
+        let new_app_addr = match new_app_address {
+            Some(address) => address - flash.as_ptr() as usize,
+            None => self.process_binaries_metadata.get().map_or(0, |metadata| {
+                metadata.new_app_address - flash.as_ptr() as usize
+            }),
+        };
+
+        // Pass the first eight bytes of the tbfheader to parse out the
+        // length of the tbf header and app. We then use those values to see
+        // if we have enough flash remaining to parse the remainder of the
+        // header.
+        let new_binary_header = match flash.get(new_app_addr..new_app_addr + 8) {
+            Some(bh) => match bh.get(0..8) {
+                Some(slice) => slice,
+                None => {
+                    return Err(ProcessBinaryError::TbfHeaderNotFound);
+                }
+            },
+            None => {
+                return Err(ProcessBinaryError::TbfHeaderNotFound);
+            }
+        };
+        let (_version, _header_length, _entry_length) =
+            match tock_tbf::parse::parse_tbf_header_lengths(
+                // new_binary_header.try_into().or(Err(ProcessBinaryError::TbfHeaderNotFound))?,
+                new_binary_header
+                    .try_into()
+                    .map_err(|_err: TryFromSliceError| ProcessBinaryError::TbfHeaderNotFound)?,
+            ) {
+                Ok((v, hl, el)) => (v, hl, el),
+                Err(tock_tbf::types::InitialTbfParseError::InvalidHeader(_entry_length)) => {
+                    // Invalid header, return Fail error. If we fail here,
+                    // let us erase the app we just wrote.
+                    return Err(ProcessBinaryError::NoBinaryFound);
+                }
+                Err(tock_tbf::types::InitialTbfParseError::UnableToParse) => {
+                    return Err(ProcessBinaryError::NoBinaryFound);
+                }
+            };
+        Ok(())
+    }
+
+    fn load_new_applications(
+        &self,
+        new_app_metadata: Option<(usize, usize)>,
+    ) -> Result<(), ProcessBinaryError> {
```

Comment:
The `new_app_metadata` shouldn't be optional. The caller should track that information.